src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,346 @@
+        let mut last_offset = None;
+        for entry in self.entries {
+            for record in build_records(entry, self.max_record_size) {
+                let kafka_record: KafkaRecord = record.try_into()?;
+                // Records of a certain region cannot be produced in parallel since their order must be static.
+                let offset = producer.produce(kafka_record).await.map(Offset).context(
+                    ProduceRecordSnafu {
+                        topic: &self.ns.topic,
+                    },
+                )?;
+                last_offset = Some(offset);
+            }
+        }
+        // Safety: there must be at least one record produced when the entries are guaranteed not empty.
+        Ok(last_offset.unwrap())
+    }
+
+    // pub(crate) async fn record(&self, record: record, producer: )
+}
+
+fn check_records(records: &[Record]) -> Result<()> {
+    let len = records.len();
+    ensure!(
+        len > 0,
+        IllegalSequenceSnafu {
+            error: "Empty sequence"
+        }
+    );
+
+    let (sequence, matched): (Vec<_>, Vec<_>) = records
+        .iter()
+        .map(|record| {
+            (
+                i32::from(record.tp),
+                record.checksum == crc32fast::hash(&record.data),
+            )
+        })
+        .unzip();
+
+    ensure!(matched.into_iter().all(|ok| ok), ChecksumMismatchedSnafu);
+
+    if len == 1 {
+        // A sequence with a single record must contain only a Full record.
+        if sequence[0] != i32::from(RecordType::Full) {
+            return IllegalSequenceSnafu {
+                error: "Missing Full record",
+            }
+            .fail();
+        }
+    } else {
+        // A sequence with multiple records must start with a First record and end with a Last record.
+        // In addition, the Middle records must in order.
+        let prefix = (0..(len - 1) as i32).collect::<Vec<_>>();
+        ensure!(
+            sequence[0..len - 1] == prefix
+                && sequence.last().unwrap() == &i32::from(RecordType::Last),
+            IllegalSequenceSnafu {
+                error: "Out of order",
+            }
+        );
+    }
+
+    Ok(())
+}
+
+fn record_type(seq: usize, num_records: usize) -> RecordType {
+    if seq == 0 {
+        RecordType::First
+    } else if seq == num_records - 1 {
+        RecordType::Last
+    } else {
+        RecordType::Middle(seq)
+    }
+}
+
+fn build_records(entry: EntryImpl, max_record_size: usize) -> Vec<Record> {
+    let payload = entry.data.len();
+    let checksum = crc32fast::hash(&entry.data);
+
+    if payload <= max_record_size {
+        let record = Record {
+            version: VERSION,
+            tp: RecordType::Full,
+            data: entry.data,
+            entry_id: entry.id,
+            ns: entry.ns,
+            checksum,
+        };
+        return vec![record];
+    }
+
+    let num_records = (payload + max_record_size - 1) / max_record_size;
```

Comment:
How about using https://doc.rust-lang.org/std/slice/struct.Chunks.html instead of manually segmenting?